sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/persistence/impl/CollectionLoggingInterceptor.java

```diff
@@ -0,0 +1,185 @@
+package uk.ac.bbsrc.tgac.miso.persistence.impl;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Queue;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import org.hibernate.EmptyInterceptor;
+import org.hibernate.type.Type;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.beans.factory.BeanFactoryAware;
+
+import com.eaglegenomics.simlims.core.User;
+
+import uk.ac.bbsrc.tgac.miso.core.data.Box;
+import uk.ac.bbsrc.tgac.miso.core.data.Boxable;
+import uk.ac.bbsrc.tgac.miso.core.data.Dilution;
+import uk.ac.bbsrc.tgac.miso.core.data.Pool;
+import uk.ac.bbsrc.tgac.miso.core.store.ChangeLogStore;
+import uk.ac.bbsrc.tgac.miso.persistence.impl.HibernateChangeLogDao.ChangeLogType;
+
+/**
+ * When “collection” entities (pools, boxes) mutate write custom change log entries
+ *
+ */
+final class CollectionLoggingInterceptor extends EmptyInterceptor implements BeanFactoryAware {
+
+  /**
+   * A change log entry. These must be written after the Hibernate session is flushed, so the required information is queued.
+   *
+   */
+  private class ChangeLogEntry {
+    long entityId;
+    ChangeLogType entityType;
+    User lastModifier;
+    String summary;
+
+    void write(ChangeLogStore changeLogStore) {
+      changeLogStore.create(entityType.name(), entityId, "contents", summary, lastModifier);
+    }
+  }
+
+  private final Queue<ChangeLogEntry> changeLogQueue = new ConcurrentLinkedQueue<>();
+
+  private ChangeLogStore changeLogStore;
+
+  private BeanFactory factory;
+
+  private String buildElementString(Set<Dilution> dilutionss) {
+    StringBuilder names = new StringBuilder();
+    for (Dilution dilution : dilutionss) {
+      if (names.length() > 0) {
+        names.append(", ");
+      }
+      names.append(dilution.getName());
+    }
+    return names.toString();
+  }
+
+  @Override
+  public boolean onFlushDirty(Object entity, Serializable id, Object[] currentState, Object[] previousState, String[] propertyNames,
```

Comment:
Could you add tests for this? The specific part I'm concerned about is how `(HashSet) x.removeAll(y)` is used to determine changes. While this should be fine, we've had several problems relating to hashcode methods and they're still not consistent across our models.